### ISSUE TITLE
System.Data.SqlClient is not supported on this platform

### ISSUE THREAD

**Issue by pabermod** [dotnet/SqlClient#1521]
Hi,

I have a library that references Microsoft.EntityFrameworkCore.SqlServer. This library has a method for migrating the database model using entity framework core.

I'm loading this library dinamically from another library with a custom AssemblyLoadContext (to avoid problems with different dependencies version and to allow to unload the dll once executed) and executing this method.

This is working with other providers I have referenced in the first library (for example Npgsql). This is also working if I copy the first library deps.json file to the deploy folder. I don't understand Why I need the deps.jsonof the library only for te Sql dependency. I also don't know how to deploy it

The exception:

System.Reflection.TargetInvocationException
  HResult=0x80131604
  Message=Exception has been thrown by the target of an invocation.
  Source=System.Private.CoreLib
  StackTrace:
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Span`1& arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at ConsoleApp10.Program.ExecuteAndUnload(AssemblyLoadContext context, String assemblyPath, WeakReference& alcWeakRef) in C:\Users\pablo.bernal\source\repos\ConsoleApp10\ConsoleApp10\Program.cs:line 205
   at ConsoleApp10.Program.Main(String[] args) in C:\Users\pablo.bernal\source\repos\ConsoleApp10\ConsoleApp10\Program.cs:line 159

  This exception was originally thrown at this call stack:
    [External Code]

Inner Exception 1:
PlatformNotSupportedException: Microsoft.Data.SqlClient is not supported on this platform.

The assembly loader
```C#
public class MyAssemblyLoadContext : AssemblyLoadContext
    {
        private AssemblyDependencyResolver _resolver;

        public MyAssemblyLoadContext(string mainAssemblyToLoadPath) : base(isCollectible: true)
        {
            _resolver = new AssemblyDependencyResolver(mainAssemblyToLoadPath);
        }

        protected override Assembly Load(AssemblyName name)
        {
            string assemblyPath = _resolver.ResolveAssemblyToPath(name);
            if (assemblyPath != null)
            {
                return LoadFromAssemblyPath(assemblyPath);
            }
            
            return null;
        }

        protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
        {
            string libraryPath = _resolver.ResolveUnmanagedDllToPath(unmanagedDllName);
            if (libraryPath != null)
            {
                return LoadUnmanagedDllFromPath(libraryPath);
            }

            return IntPtr.Zero;
        }
    }
```

My code to execute the method
```cs
var SourceDirectory = @"C:\dev\myproject\bin\net6.0";
            string ConnectionString = @"Data Source=myhost\SQLExpress;Initial Catalog=mydb;User Id=myuser;Password=mypass;";
            string ProviderName = "System.Data.SqlClient";

            var assembly = context.LoadFromAssemblyPath(assemblyPath);
            alcWeakRef = new WeakReference(context, trackResurrection: true);

            var adContext = assembly.GetType("MyType");
            var instance = Activator.CreateInstance(adContext, ConnectionString, ProviderName, null);
            PropertyInfo databaseProperty = adContext.GetProperty("Database");
            var database = databaseProperty.GetValue(instance);

            var efAssembly = context.LoadFromAssemblyPath(Path.Combine(SourceDirectory, "Microsoft.EntityFrameworkCore.Relational.dll"));
            var extensionsType = efAssembly.GetType("Microsoft.EntityFrameworkCore.RelationalDatabaseFacadeExtensions");
            MethodInfo mmigrateMethod = extensionsType.GetMethod("Migrate");

            mmigrateMethod.Invoke(database, new object[] { database });
            context.Unload();
```

EF Core version: 6.0.2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows 10
IDE: Visual Studio 2022 (latest version)

Packages used in called library:

	<ItemGroup>
		<PackageReference Include="Microsoft.Data.SqlClient" Version="4.1.0" />
		<PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.2" />
		<PackageReference Include="Microsoft.EntityFrameworkCore.InMemory" Version="6.0.2" />
		<PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="6.0.2" />
		<PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.2" />
		<PackageReference Include="Devart.Data.Oracle.EFCore" Version="9.16.1434" />
		<PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="6.0.3" />
		<PackageReference Include="Oracle.EntityFrameworkCore" Version="6.21.5" />
		<PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="6.0.1" />
	</ItemGroup>

**Comment by JRahnama** [comment]
@pabermod what is variable `context` in the above code? Is it `AssemblyLoadContext`?

**Comment by pabermod** [comment]
> @pabermod what is variable `context` in the above code? Is it `AssemblyLoadContext`?

Yes sorry it is an instance of MyAssemblyLoadContext

var context = new MyAssemblyLoadContext(assemblyPath) and assemblyPath is the assembly full path being loaded

**Comment by JRahnama** [comment]
ok, so the `assemblyPath` is pointing to `Microsoft.Data.SqlClient.dll` if I am not mistaken? have you picked it from the correct runtime folder?

one more thing, the title is mentioning `System.Data.SqlClient` and the code is from `Microsoft.Data.SqlClient`. I think you are using `Microsoft.Data.SqlClient` and the title was just a typo? Am I correct?

**Comment by JRahnama** [comment]
I think it works when you add `deps.json` because it addresses the correct path for runtime dll. You can look into that folder and see the correct path of it

**Comment by pabermod** [comment]
@JRahnama assemblyPath is my dll that references Microsoft.Data.SqlClient. While debugging I can see that context has apparently loaded the dependencias.
The data provider name is still System.Data.Sql even the package being Microsoft.Data.SqlClient.

This library is working correctly if I load throught project reference but I need to load it on runtime

**Comment by JRahnama** [comment]
@pabermod When you reference it by project reference the correct runtime is picked up. Look at below image: 
![image](https://user-images.githubusercontent.com/32077152/155229607-00df9517-45a8-41bb-9512-f3c682bb5df8.png)

can you, if it is not done yet, try referencing M.D.SqlClient from the runtime folder?

**Comment by pabermod** [comment]
I Will try it and let you know

**Comment by pabermod** [comment]
@JRahnama If I specifically load the Microsoft.Data.SqlClient from runtimes\win\lib\netcoreapp3.1 it is working... But this doesn't seem to be a good approach for solving this. Is there a better a solution?

**Comment by pabermod** [comment]
Update: it is not working, now it throws an error: Unable to load DLL 'Microsoft.Data.SqlClient.SNI.dll'

**Comment by JRahnama** [comment]
@pabermod 
Please note, the lib\ folder DLLs are fallback not-supported DLLs for .NET Core and .NET Standard whereas the runtime DLLs are the official platform compatible DLLs that one should be able to work with.

for a better understanding you can take a look at #1347

**Comment by pabermod** [comment]
Thanks!

I finally managed to get it working by manually specifying the runtimes for the current RID changing `MyAssemblyLoadContext`:

```cs
    public class MyAssemblyLoadContext : AssemblyLoadContext
    {
        private AssemblyDependencyResolver _resolver;
        private string _baseDirectory;

        public MyAssemblyLoadContext(string mainAssemblyToLoadPath) : base(isCollectible: true)
        {
            _resolver = new AssemblyDependencyResolver(mainAssemblyToLoadPath);
            _baseDirectory = Path.GetDirectoryName(mainAssemblyToLoadPath);
        }

        protected override Assembly Load(AssemblyName name)
        {
            string assemblyPath = _resolver.ResolveAssemblyToPath(name);
            if (assemblyPath != null)
            {
                return LoadFromAssemblyPath(assemblyPath);
            }

            return null;
        }

        protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
        {
            string libraryPath = ResolveUnmanagedDllPath(unmanagedDllName);
            if (libraryPath != null)
            {
                return LoadUnmanagedDllFromPath(libraryPath);
            }

            return IntPtr.Zero;
        }

        private string ResolveUnmanagedDllPath(string unmanagedDllName)
        {
            string dllPath = _resolver.ResolveUnmanagedDllToPath(unmanagedDllName);
            if (dllPath != null)
                return dllPath;
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                dllPath = Path.Combine(_baseDirectory, "runtimes", $"win-{RuntimeInformation.ProcessArchitecture.ToString().ToLowerInvariant()}", "native", unmanagedDllName);
                if (File.Exists(dllPath))
                    return dllPath;
            }
            return null;
        }
    }
```

I think we can close this issue as it doesn't seem to be a bug... Thanks for your help

**Comment by JRahnama** [comment]
@pabermod glad we could help. Closing the issue as it is resolved.

**Comment by itadapter** [comment]
We spent the whole day trying to understand the same issue.
Not sure why this issue is closed but the package does not work properly.

Confusion started back in 2018 with `System.Data.SqlClient.dll` and continues on with `Microsoft.Data.SqlClient.5.0` package to this day (Sep 2022).

**The build system is not putting proper DLL in the \bin folder**. Why? We are targeting Net6.0 and upgraded to Microsoft* pkg hoping this issue would have been addressed by now. It is not addressed.

The instruction here: https://github.com/dotnet/SqlClient/wiki/Frequently-Asked-Questions#11-why-do-i-get-a-platformnotsupported-exception-when-my-application-hits-a-sqlclient-method

says to *"Ultimately, you want to use the NuGet package reference infrastructure or you would have to implement all this target framework and platform support logic yourself when determining which DLL to load."* - we are using Nuget reference and standard dotnet build, yet the invalid file is copied into /bin even on WIndowsx64 runtime.

We have no choice but include binary dll files to COPY-OVER the one that is being created by MSBUILD using a surrogated file copy step. This is very very bad practice and we still do not understand why this fails for flagship MSFT data access technology out of the box?  The custom assembly load context approach above is a hack.... why does this not work in .NET6 out of the box?
